File: Java/src/main/java/com/gildedrose/GenericItem.java
package com.gildedrose;

/**
 *
 * @author markd
 */
public class GenericItem extends Item{
    
    QualityRating qualityRating;
    int QUALITY_CHANGE_BEFORE_SELLBY;
    int QUALITY_CHANGE_AFTER_SELLBY;
    
    public GenericItem(String name, int sellIn, int quality) {
        super(name, sellIn, quality);
        qualityRating = new QualityRating(quality);
        QUALITY_CHANGE_BEFORE_SELLBY = -1;
        QUALITY_CHANGE_AFTER_SELLBY = -2;
    }
    
    public void updateQuality(){
        if(sellIn > 0) {
            qualityRating.modify(QUALITY_CHANGE_BEFORE_SELLBY);
        }
        else{
            qualityRating.modify(QUALITY_CHANGE_AFTER_SELLBY);
        }
        quality = qualityRating.get();
        sellIn--;
    }
    
}

File: ruby/generic_item.rb
class GenericItem

  def initialize(item)
    @item = item
    @quality_rating = QualityRating.new(item.quality)
    @quality_change_before_sellby = -1
    @quality_change_after_sellby = -2
  end

  def update_quality
    if @item.sell_in.positive?
      @quality_rating.modify(@quality_change_before_sellby)
    else
      @quality_rating.modify(@quality_change_after_sellby)
    end
    @item.quality = @quality_rating.get
    --@item.sell_in
  end

  def item
    @item
  end
end